server/ec2spotmanager/CloudProvider/EC2SpotCloudProvider.py

```diff
@@ -0,0 +1,379 @@
+                                        result.state)
+                else:  # state=failed
+                    msg = "Request %s is %s and %s." % (req_id, result.status.code, result.state)
+                    self.logger.error("[Pool %d] %s", pool_id, msg)
+                    failed_requests[req_id] = {}
+                    failed_requests[req_id]['action'] = 'disable_pool'
+                    failed_requests[req_id]['pool'] = pool_id
+                    break
+            elif result is None:
+                self.logger.info("[Pool %d] spot request %s is still open", pool_id, req_id)
+            else:
+                self.logger.warning("[Pool %d] spot request %s returned %s", pool_id, req_id, type(result).__name__)
+
+        return (successful_requests, failed_requests)
+
+    def check_instances_state(self, pool_id, region):
+
+        instance_states = {}
+        cluster = EC2Manager(None)
+        try:
+            cluster.connect(region=region, aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
+                            aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY)
+        except Exception as msg:
+            raise Exception({'type': 'unclassified', 'data': msg})
+
+        boto_instances = cluster.find(filters={"tag:" + SPOTMGR_TAG + "-PoolId": str(pool_id)})
+
+        for instance in boto_instances:
+            if instance.state_code not in [INSTANCE_STATE['shutting-down'], INSTANCE_STATE['terminated']]:
+                instance_states[instance.id] = {}
+                instance_states[instance.id]['status'] = instance.state_code & 255
+                instance_states[instance.id]['tags'] = instance.tags
+
+        return instance_states
+
+    def get_spot_prices(self, regions, instance_types=None, use_multiprocess=False):
+        if use_multiprocess:
+            from multiprocessing import Pool, cpu_count
+            pool = Pool(cpu_count())
+
+        try:
+            results = []
+            for region in regions:
+                args = [region, instance_types]
+                if use_multiprocess:
+                    results.append(pool.apply_async(self._get_spot_price_per_region, args))
+                else:
+                    results.append(self._get_spot_price_per_region(*args))
+
+            prices = {}
+            for result in results:
+                if use_multiprocess:
+                    result = result.get()
+                for instance_type in result:
+                    prices.setdefault(instance_type, {})
+                    prices[instance_type].update(result[instance_type])
+        finally:
+            if use_multiprocess:
+                pool.close()
+                pool.join()
+
+        return prices
+
+    def get_price_data(self, config):
+        price_data = {}
+        for instance_type in config.ec2_instance_types:
+            price_data[instance_type] = "EC2Spot:price:" + instance_type
+        return price_data
+
+    def get_image_names(self, config):
```

Comment:
As mentioned in `CloudProvider.py`, this function does not belong to the interface.